sched/adaptdl_sched/policy/pollux.py

```diff
@@ -127,19 +127,31 @@ def optimize(self, jobs, nodes, base_allocations, node_template):
             dict: map from job keys to their optimized resource allocations,
                 in the form of a list of a node key for each replica.
         """
-        jobs = OrderedDict(  # Sort jobs in FIFO order.
-            sorted(jobs.items(), key=lambda kv: kv[1].creation_timestamp))
+
+        # Consider job preemptible only if it already has an allocation
+        def isnonpreemptible(key, job):
+            return not job.preemptible and \
+                    base_allocations.get(key, []) != []
```

Comment:
I think the logic around when a job should be re-allocated should cover more cases. For example:

- What if one of a non-preemptible job's replicas gets evicted due to some reason outside of AdaptDL's control? Need to have some defined behavior here and confirm it works.
- What if the allocator sets some resources for a non-premptible job, but before the controller has a chance to create those pods, the resources are taken away by some non-AdaptDL pods? In this scenario, the job hasn't started and it's not really something the user did wrong. So I presume we'd want the allocator to keep proposing new allocations until it works?

Reply:
1. For eviction outside of ADL's control, I think we should just fail the NP job. This can be handled in the controller.

2. The allocator will keep proposing new allocations because the job won't be considered pinned because in your scenario controller will empty the allocation for it, which will be passed through `prev_allocations` to the next call to the allocator. So the job will still be NP but not pinned and qualify for reallocation.

Reply:
1. SGTM
2. `in your scenario controller will empty the allocation for it` is that actually true? I can't find where the controller does that..

Reply:
1 is handled now. As per offline discussion, we will have to handle 2 through a custom pod scheduler.

Reply:
Can you help create an issue to track this?

Reply:
created issue #42